src/texture_cache.rs

```diff
@@ -799,100 +799,86 @@ impl TextureCache {
+
+                    let cache_item = TextureCacheItem::new(page.texture_id,
+                                                           user_x0, user_y0,
+                                                           allocated_rect,
+                                                           requested_rect,
+                                                           &Size2D::new(page.texture_size, page.texture_size));
+                    *self.items.get_mut(image_id) = cache_item;
+
+                    return AllocationResult {
+                        item: self.items.get(image_id).clone(),
+                        kind: AllocationKind::TexturePage,
                     }
-                    let free_texture_level = free_texture_levels.pop().unwrap();
-                    let texture_id = free_texture_level.texture_id;
-  
-                    let page = TexturePage::new(texture_id, texture_size);
-                    page_list.push(page);
                 }
-
-                match page_list.last_mut().unwrap().allocate(&allocation_size, filter) {
-                    Some(location) => location,
-                    None => {
-                        // Fall back to standalone texture allocation.
-                        let texture_id = self.free_texture_ids
-                                             .pop()
-                                             .expect("TODO: Handle running out of texture ids!");
-                        let cache_item = TextureCacheItem::new(
-                            texture_id,
-                            user_x0, user_y0,
-                            Rect::new(Point2D::zero(), requested_size),
-                            Rect::new(Point2D::zero(), requested_size),
-                            &requested_size);
-                        *self.items.get_mut(image_id) = cache_item;
-
-                        return AllocationResult {
-                            item: self.items.get(image_id).clone(),
-                            kind: AllocationKind::Standalone,
+                None => {
+                    if !page_list.is_empty() && page_list.last().unwrap().can_grow() {
+                        let last_page = page_list.last_mut().unwrap();
+                        // Grow the texture.
+                        let texture_size = cmp::min(last_page.texture_size * 2,
+                                                    max_texture_size());
+                        self.pending_updates.push(TextureUpdate {
+                            id: last_page.texture_id,
+                            op: texture_grow_op(texture_size, format, mode),
+                        });
+                        last_page.grow(texture_size);
+
+                        self.items.for_each_item(|item| {
+                            if item.texture_id == last_page.texture_id {
+                                item.texture_size = Size2D::new(texture_size, texture_size);
+                            }
+                        });
```

Comment:
nit: you could add `continue` here and then eliminate even more rightward drift by removing the `else` clause :)